presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/TransformSpatialPredicates.java

```diff
@@ -110,50 +114,117 @@
+        public Result apply(FilterNode node, Captures captures, Context context)
+        {
+            JoinNode joinNode = captures.get(JOIN);
+
+            Expression filter = node.getPredicate();
+            List<FunctionCall> spatialFunctions = extractSupportedSpatialFunctions(filter);
+
+            for (FunctionCall spatialFunction : spatialFunctions) {
+                Result result = tryCreateSpatialJoin(context, joinNode, filter, node.getId(), node.getOutputSymbols(), spatialFunction, metadata);
+                if (!result.isEmpty()) {
+                    return result;
+                }
+            }
+
+            List<ComparisonExpression> spatialComparisons = extractSupportedSpatialComparisons(filter);
+            for (ComparisonExpression spatialComparison : spatialComparisons) {
+                Result result = tryCreateSpatialJoin(context, joinNode, filter, node.getId(), node.getOutputSymbols(), spatialComparison, metadata);
+                if (!result.isEmpty()) {
+                    return result;
+                }
+            }
+
+            return Result.empty();
+        }
     }
 
-    @Override
-    public Result apply(FilterNode node, Captures captures, Context context)
+    public static final class TransformSpatialPredicateToLeftJoin
+            implements Rule<JoinNode>
     {
-        JoinNode joinNode = captures.get(JOIN);
+        private static final Pattern<JoinNode> PATTERN = join().matching(node -> node.getCriteria().isEmpty() && node.getFilter().isPresent() && node.getType() == LEFT && !node.isSpatialJoin());
 
-        Expression filter = node.getPredicate();
-        List<FunctionCall> spatialFunctions = extractSupportedSpatialFunctions(filter);
+        private final Metadata metadata;
 
-        for (FunctionCall spatialFunction : spatialFunctions) {
-            Result result = tryCreateSpatialJoin(context, node, joinNode, filter, spatialFunction);
-            if (!result.isEmpty()) {
-                return result;
-            }
+        public TransformSpatialPredicateToLeftJoin(Metadata metadata)
+        {
+            this.metadata = metadata;
         }
 
-        List<ComparisonExpression> spatialComparisons = extractSupportedSpatialComparisons(filter);
-        for (ComparisonExpression spatialComparison : spatialComparisons) {
-            Result result = tryCreateSpatialJoin(context, node, joinNode, filter, spatialComparison);
-            if (!result.isEmpty()) {
-                return result;
-            }
+        @Override
+        public boolean isEnabled(Session session)
+        {
+            return isSpatialJoinEnabled(session);
+        }
+
+        @Override
+        public Pattern<JoinNode> getPattern()
+        {
+            return PATTERN;
         }
 
-        return Result.empty();
+        @Override
+        public Result apply(JoinNode joinNode, Captures captures, Context context)
+        {
+            Expression filter = joinNode.getFilter().get();
```

Comment:
@arhimondr `PredicatePushDown` converts an inner join without equi-criteria into a filter + cross-join (because nested loop join doesn't support filters and projections). Hence, `TransformSpatialPredicateToJoin` needs to run *after* last run of the `PredicatePushDown` and needs to handle filter + cross-join. `PredicatePushDown` doesn't do that for the left join (because left join with no equi-criteria is executed as a hash join).